packages/mangrove-arbitrage/src/MgvArbitrage.sol

```diff
@@ -75,176 +111,132 @@ contract MgvArbitrage is AccessControlled {
-      ISwapRouter.ExactInputSingleParams memory exchangeParams = ISwapRouter.ExactInputSingleParams({
-        tokenIn: params.takerGivesToken,
-        tokenOut: token,
-        fee: fee,
-        recipient: address(this),
-        deadline: type(uint).max,
-        amountIn: amountOut,
-        amountOutMinimum: 0,
-        sqrtPriceLimitX96: 0
-      });
-      router.exactInputSingle(exchangeParams);
+    bool zeroForOne = givesToken < wantsToken; // tokenIn < tokenOut
+    (int amount0, int amount1) = pool.swap(
+      address(this), zeroForOne, amount, zeroForOne ? MIN_SQRT_RATIO + 1 : MAX_SQRT_RATIO - 1, abi.encode(givesToken)
+    );
+
+    if (amount0 > 0) {
+      amountGives = uint(amount0);
+    } else {
+      amountWants = uint(-amount0);
     }
-  }
 
-  /// @notice this exchanges, on Mangrove, from the token the contract gain from the arbitrage, to the token the contract is holding
-  /// @param params The parameters needed to do the arbitrage
-  /// @param token the token the contract is holding
-  /// @param amountOut The amount received from the arbitrage
-  function postExchangeOnMgv(ArbParams calldata params, address token, uint amountOut) internal {
-    if (token != address(0) && token != params.takerGivesToken) {
-      (, uint takerGave,,) = mgv.marketOrder({
-        outbound_tkn: token,
-        inbound_tkn: params.takerGivesToken,
-        takerWants: 0,
-        takerGives: amountOut,
-        fillWants: false
-      });
-      require(takerGave == amountOut, "MgvArbitrage/notEnoughOnMgv");
+    if (amount1 > 0) {
+      amountGives = uint(amount1);
+    } else {
+      amountWants = uint(-amount1);
     }
   }
 
-  /// @notice this exchanges, on Uniswap, from the token the contract is holding, to the token the contract needs for the arbitrage
-  /// @param params The parameters needed to do the arbitrage
-  /// @param token the token the contract is holding
-  /// @param fee the fee used for the swap on Uniswap
-  function preExchangeOnUniswap(ArbParams calldata params, address token, uint24 fee) internal {
-    if (token != address(0) && token != params.takerGivesToken) {
-      ISwapRouter.ExactOutputSingleParams memory exchangeParams = ISwapRouter.ExactOutputSingleParams({
-        tokenIn: token,
-        tokenOut: params.takerGivesToken,
-        fee: fee,
-        recipient: address(this),
-        deadline: type(uint).max,
-        amountOut: params.takerGives,
-        amountInMaximum: IERC20(token).balanceOf(address(this)),
-        sqrtPriceLimitX96: 0
-      });
-      router.exactOutputSingle(exchangeParams);
-    }
-  }
+  /// @inheritdoc IUniswapV3SwapCallback
+  function uniswapV3SwapCallback(int amount0Delta, int amount1Delta, bytes calldata data) external {
+    require(pools[msg.sender] == true); // ensure only approved UniswapV3 can call this function
+    require(isArbitraging == true);
+    isArbitraging = false;
+    // TODO: maybe we should check that we first called doArbitrage fn ?
```

Comment:
comment is no longer relevant provided `isArbitraging` can only be set in `doArbitrage...`